### PR TITLE
Remove setuptools dependencie, improve release process

### LINKED ISSUE #1016: astroid has an undeclared dependency on setuptools.
The dependency is here: https://github.com/PyCQA/astroid/blob/1342591e2beb955a377e4486e5595478f79789e8/astroid/__pkginfo__.py#L29

The lack of declaration is here: https://github.com/PyCQA/astroid/blob/1342591e2beb955a377e4486e5595478f79789e8/setup.cfg#L37-L41

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
## Description

This MR replace setuptools_scm by tbump and a small script. This should permit to remove the setuptools hidden dependencie and facilitate the release process. Next step is to add the release itself in github actions.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |
| ✓   | :hammer: Refactoring   |

## Related Issue

Closes #1016

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ install:
   - python -m pip --version
   - python -m tox --version
 script:
-  - python -m pip install .
+  - python -m pip install . --use-feature=in-tree-build
```
Comment (cdce8p) [inline .travis.yml:16]:
```suggestion
  - python -m pip install -e .
```
That would probably do the trick. I'm just not sure it's at all necessary. With `deploy` gone from travis. There is no need to do anything there.

Ideally I would suggest to remove it completely. Same with `appveyor`.

File: doc/release.md [hunk 2]
```diff
@@ -4,27 +4,13 @@ So, you want to release the `X.Y.Z` version of astroid ?
...
-4. Submit your changes in a merge request and make sure the tests are passing on
-   Travis/GithubActions: https://travis-ci.org/PyCQA/astroid/
-
-5. Do the actual release by tagging the master with `vX.Y.Z` (ie `v1.6.12` or `v3.0.0a0`
-   for example).
-
-Until the release is done via Travis or GitHub actions on tag, run the following
-commands:
+5. Until the release is done via Travis or GitHub actions on tag, run the following
+   commands:
```
Comment (cdce8p) [inline doc/release.md:13]: If you like, take a look at this script here: [cdce8p/python-typing-update -> release.yaml]( https://github.com/cdce8p/python-typing-update/blob/f79966eb104f220fdcc78f87764eb74ab6e527c2/.github/workflows/release.yaml). It's the release workflow for one of my personal projects. That should also work for pylint / astroid. The only requirement is that `PYPI_API_TOKEN` is added as secret.

File: script/bump_changelog.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+def main() -> None:
+    args = parse_args()
+    if "dev" not in args.version:
+        run(args.version)
+
+
+def parse_args() -> argparse.Namespace:
+    parser = argparse.ArgumentParser(add_help=__doc__)
+    parser.add_argument("version", help="The version we want to release")
+    return parser.parse_args()
```
Comment (cdce8p) [inline script/bump_changelog.py:32]: For such a small script, I would just do the argument parsing in `main`.

**Review by cdce8p — COMMENTED** [review]
Seems like `tbump` might actually work 😉 I left some comments. This one does replace #1018, right?

File: script/bump_changelog.py (hunk 3 above)
Comment (Pierre-Sassoulas) [reply script/bump_changelog.py:32]: Done :)

File: .travis.yml (hunk 1 above)
Comment (Pierre-Sassoulas) [reply .travis.yml:16]: Yeah, nice catch, sadly I don't have the right on Travis either. At least it does not fail when doing nothing...

File: doc/release.md (hunk 2 above)
Comment (Pierre-Sassoulas) [reply doc/release.md]: I'm going to try to release a dev version in order to test that. I don't want to be surprised when releasing ``2.6.0`` 😄

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+name: Release
+
+on:
+  release:
+    types:
+      - published
```
Comment (cdce8p) [inline .github/workflows/release.yml:6]: Just a note, currently the workflow is run when a new `release` is published.
https://docs.github.com/en/actions/reference/events-that-trigger-workflows

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+          python-version: ${{ env.DEFAULT_PYTHON }}
+      - name: Install requirements
+        run: |
+          python -m pip install -U pip twine wheel
+          python -m pip install -U "setuptools>=56.0.0"
+      - name: Build distributions
+        run: |
+          python setup.py sdist bdist_wheel
+      - name: Upload to PyPI
+        if: github.event_name == 'release' && startsWith(github.ref, 'refs/tags')
```
Comment (cdce8p) [inline .github/workflows/release.yml:31]: Also note this line here

Comment (Pierre-Sassoulas) [reply .github/workflows/release.yml:31]: The release on github should start with a tag ? For example I pushed ``2.6.0-dev0`` as a tag right now, do I need to release a version, or is there something else ? (This is a nice workflow, it forces us to have a release and a tag)

Comment (cdce8p) [reply .github/workflows/release.yml:31]: > For example I pushed 2.6.0-dev0 as a tag right now, do I need to release a version

You do. The release just needs to for the same commit you tagged. I usually create the tag together with the release, but it should work either way.

Comment (cdce8p) [reply .github/workflows/release.yml:31]: I believe it shouldn't matter if it's a pre-release.

Comment (cdce8p) [reply .github/workflows/release.yml:31]: @Pierre-Sassoulas Might be that you need to add the workflow to the master branch first as it isn't trigged on a PR action.

Comment (Pierre-Sassoulas) [reply .github/workflows/release.yml:31]: OK I'm going to try that and upgrade the release process.

**Comment by Pierre-Sassoulas** [comment]
Next step, same thing in pylint ! :)

**Pull request merged**